File: java/sdk/sdk-core/src/main/java/com/novax/sdk/core/http/Interceptor.java
package com.novax.sdk.core.http;

import java.io.IOException;

/**
 * Cross-cutting hook for outbound requests / inbound responses. All such
 * concerns — auth, retry, logging, header injection — live as interceptors,
 * not as ad-hoc logic in the client.
 *
 * <p>Implementations <em>must</em> call {@code chain.proceed(...)} exactly once
 * (or short-circuit by building a {@link SdkResponse} themselves).
 */
public interface Interceptor {

    SdkResponse intercept(Chain chain) throws IOException;

    interface Chain {
        SdkRequest request();

        SdkResponse proceed(SdkRequest request) throws IOException;
    }
}

File: python/novax_sdk/http/interceptor.py
from __future__ import annotations

from abc import ABC, abstractmethod

from .sdk_request import SdkRequest
from .sdk_response import SdkResponse


class Chain(ABC):
    @property
    @abstractmethod
    def request(self) -> SdkRequest:
        ...

    @abstractmethod
    def proceed(self, request: SdkRequest) -> SdkResponse:
        ...


class Interceptor(ABC):
    """Cross-cutting hook. Implementations must call chain.proceed() exactly once."""

    @abstractmethod
    def intercept(self, chain: Chain) -> SdkResponse:
        ...
